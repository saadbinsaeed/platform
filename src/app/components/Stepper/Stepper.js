/* @flow */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Stepper from 'app/components/atoms/Stepper/Stepper';
import { connect } from 'react-redux';
import { showStepperSave, toggleAppHeader } from 'store/actions/app/appActions';
import Form from 'app/components/atoms/Form/Form';
import history from 'store/History';
import { bind } from 'app/utils/decorators/decoratorUtils';

/**
 *
 */
export class StepperContainer extends PureComponent<Object, Object> {
    static propTypes = {
        steps: PropTypes.arrayOf(
            PropTypes.shape({
                title: PropTypes.string.isRequired,
                subTitle: PropTypes.string,
                formId: PropTypes.string
            })
        ).isRequired,
        onClose: PropTypes.func,
        onDone: PropTypes.func.isRequired,
        toggleAppHeader: PropTypes.func.isRequired,
        showStepperSave: PropTypes.func.isRequired
    };

    static defaultProps = {};

    constructor(props: Object) {
        super(props);
        this.state = {
            step: 1,
            disabled: false
        };
        this.props.toggleAppHeader(true);
        this.props.showStepperSave();
    }

    componentWillUnmount() {
        this.props.toggleAppHeader(false);
    }

    @bind
    mbHandleError(mbError: any) {
        this.setState({ disabled: false });
        if (mbError instanceof Error) {
            // if there was error navigate back to the last step
            return this.onPrevious();
        }
        this.onClose();
    };

    @bind
    onNext() {
        const { step } = this.state;
        step <= this.props.steps.length && this.setState(prevState => ({ step: prevState.step + 1 }));
    }

    @bind
    onPrevious() {
        const { step } = this.state;
        step > 1 && this.setState(prevState => ({ step: prevState.step - 1 }));
    }

    @bind
    onDone() {
        const { onDone } = this.props;
        if (onDone) {
            const res = this.props.onDone();
            this.setState({ disabled: true });
            if (res.then && typeof res.then === 'function') {
                res.then(this.mbHandleError, this.mbHandleError);
            } else {
                this.onClose();
            }
        }
    }

    @bind
    onClose() {
        if (this.props.onClose) {
            this.props.onClose();
        } else {
            history.pushBack();
        }
    }

    @bind
    onFormSubmit(event: Object) {
        event.preventDefault();
        if (this.state.step === this.props.steps.length) {
            this.onDone();
        } else {
            this.onNext();
        }
    }

    render() {
        const { step, disabled } = this.state;
        const { config } = this.props;
        const { title, subTitle, formId, content } = this.props.steps[step - 1] || {};
        return (
            <Stepper
                title={step > this.props.steps.length ? 'Loading' : title}
                subTitle={step > this.props.steps.length ? '' : subTitle}
                step={step}
                steps={this.props.steps && this.props.steps.length}
                onNext={this.onNext}
                onPrevious={this.onPrevious}
                onClose={this.onClose}
                formId={formId}
                disabled={disabled}
                hideOnSave={config.hideOnSave}
            >
                <Form onSubmit={this.onFormSubmit} id={formId}>
                    {content}
                </Form>
            </Stepper>
        );
    }
}

export default connect(
    state => ({ config: state.app.stepper }),
    {
        toggleAppHeader,
        showStepperSave
    }
)(StepperContainer);
