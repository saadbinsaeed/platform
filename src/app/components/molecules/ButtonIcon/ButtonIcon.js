/* @flow */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Button from 'app/components/atoms/Button/Button';
import Icon from 'app/components/atoms/Icon/Icon';
import Text from 'app/components/atoms/Text/Text';
import { IconTypeProps, SizeProps } from 'app/utils/propTypes/common';

const ButtonIconStyle = styled(Button)`
    font-size: ${({ size, theme }) => theme.iconSize[size]};
    padding: 0.199em 0.35em;
    box-shadow: none;
    ${({ paddind }) => paddind ? 'padding: .1rem;' : ''};
    ${({ fluid }) => fluid ? 'width: 100%;' : ''};
    ${({ rounded }) => rounded ? 'border-radius: 500rem;' : ''};
    ${({ theme, backgroundColor }) => backgroundColor && theme ? `background: ${theme.color[backgroundColor]}; box-shadow: ${theme.shadow.z1}; ` : ''}
    & .Icon {
        margin: 0;
    }
`;

const ButtonIconLabel = styled(Text)`
    margin-left: 1rem;
    font-size: 1rem;
    color: white;
    font-weight: 500;
    text-decoration: none;
    text-align: center;
`;

const ButtonIcon = ({ padding, title, type, icon, size, backgroundColor, iconColor, rounded, onClick, loading, label, buttonType, ...rest }: Object) => {
    return (
        <Fragment>
            <ButtonIconStyle type={buttonType} color={backgroundColor} onClick={loading ? null : onClick} size={size} loading={loading}
                rounded={rounded} title={title} {...rest}>
                <Icon type={type} name={icon} color={iconColor} size={size} />
                {label && (<ButtonIconLabel paddind>{label}</ButtonIconLabel>)}
            </ButtonIconStyle>
        </Fragment>
    );
};

ButtonIcon.propTypes = {
    title: PropTypes.string,
    type: IconTypeProps,
    icon: PropTypes.string,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    backgroundColor: PropTypes.string,
    iconColor: PropTypes.string,
    size: SizeProps,
    loading: PropTypes.bool,
    rounded: PropTypes.bool,
    onClick: PropTypes.func,
    className: PropTypes.any,
    buttonType: PropTypes.string,
};
ButtonIcon.defaultProps = {
    rounded: true,
    buttonType: 'button',

};

export default ButtonIcon;
