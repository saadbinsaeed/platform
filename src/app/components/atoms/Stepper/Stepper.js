/* @flow */

import React from 'react';
import PropTypes from 'prop-types';
import ButtonIcon from 'app/components/molecules/ButtonIcon/ButtonIcon';
import styled, { css } from 'styled-components';
import Button from 'app/components/atoms/Button/Button';
import Icon from 'app/components/atoms/Icon/Icon';
import { Row, Col } from 'react-styled-flexboxgrid';
import Title from 'app/components/atoms/Title/Title';

const commonStyles = css`
    white-space: nowrap;
    max-width: 300px;
    overflow: hidden;
`;

const HeaderSubTitle = styled(Title)`
    color: rgba(255, 255, 255, 0.75);
    font-size: 0.7rem;
    font-weight: 300;
    min-height: 16px;
    ${commonStyles}
`;
HeaderSubTitle.displayName = 'HeaderSubTitle';

const HeaderTitle = styled(Title)`
    ${commonStyles}
`;
HeaderTitle.displayName = 'HeaderTitle';

const StyledRow = styled(Row)`
    background: ${({ color, theme }) => color || theme.header.background};
`;

const ButtonStyle = styled(Button)`
    font-size: 16px;
    height: 50px;
    &:disabled {
        background-color: #2890d9 !important;
    }
`;

ButtonStyle.displayName = 'ButtonStyle';

const NextButton = ({ title, subTitle, children, step, steps, onPrevious, onClose, formId, disabled, hideOnSave }: Object) => {
    if (step < steps) {
        return (
            <ButtonStyle title="Go to the next step" type="submit" form={formId}>
                Next
                <Icon style={{ marginLeft: '10px' }} name="chevron-right" size="lg" />
            </ButtonStyle>
        );
    }
    const doneProps = {
        title: 'Click to complete',
        type: 'submit',
        form: formId,
        disabled,
        style: { marginLeft: '15px' }
    };
    if (step > 1) {
        return <ButtonStyle {...doneProps}>Done</ButtonStyle>;
    }
    if (!hideOnSave) {
        return <ButtonStyle {...doneProps}>Save</ButtonStyle>;
    }
    return null;
};

const StepperComponent = (props: Object) => {
    const { title, subTitle, children, step, steps, onPrevious, onClose } = props;
    return (
        <div>
            <StyledRow middle="xs">
                <Col xs={1}>
                    <ButtonIcon icon="close" title="Close Current View" onClick={onClose} />
                </Col>
                <Col xs={11}>
                    <HeaderTitle as="h1">{title}</HeaderTitle>
                    <HeaderSubTitle as="h2">{subTitle}</HeaderSubTitle>
                </Col>
            </StyledRow>

            <StyledRow middle="xs">
                <Col xs={1}>{step > 1 ? <ButtonIcon icon="chevron-left" title="Go to the previous step" onClick={onPrevious} /> : null}</Col>
                <Col xs={8} md={10}>
                    <span className="step-status">
                        STEP {step} / {steps}
                    </span>
                </Col>
                <Col xs={3} sm={2} md={1}>
                    <NextButton {...props} />
                </Col>
            </StyledRow>
            <div className="content-wrap">{children}</div>
        </div>
    );
};

StepperComponent.propTypes = {
    title: PropTypes.string.isRequired,
    step: PropTypes.number.isRequired,
    steps: PropTypes.number.isRequired,
    onClose: PropTypes.func.isRequired,
    children: PropTypes.element.isRequired,
    subTitle: PropTypes.string
};

StepperComponent.defaultProps = {
    subTitle: ''
};

export default StepperComponent;
