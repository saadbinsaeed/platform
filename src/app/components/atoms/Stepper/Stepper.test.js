import React from 'react';
import Stepper from 'app/components/atoms/Stepper/Stepper';
import {shallow} from 'enzyme';
import theme from 'app/themes/theme.default';
import ButtonIcon from 'app/components/molecules/ButtonIcon/ButtonIcon';
import {ThemeProvider} from 'styled-components';
import 'jest-styled-components';
import context from 'jest-plugin-context';

// import Button from 'app/components/atoms/Button/Button';
// import renderer from 'react-test-renderer'


const props = {
    title: 'ABC',
    subTitle: 'XYZ',
    children: <div/>,
    step: 2,
    steps: 3,
    onClose: () => {
    },
    onStepChange: () => {
    }
};

describe('<Stepper />', () => {
    const shallowWithTheme = (tree, theme) => {
        const context = shallow(<ThemeProvider theme={theme}/>)
            .instance()
            .getChildContext();
        return shallow(tree, {context});
    };

    describe('title', () => {
        it('should render title in Header Title Element', () => {
            const wrapper = shallow(<Stepper {...props} title="test title"/>);
            expect(wrapper.find('HeaderTitle').children().text()).toEqual('test title');
        });
        it('should render title with h1 as a "as" prop ', () => {
            const wrapper = shallow(<Stepper {...props} title="test title"/>);
            expect(wrapper.find('HeaderTitle').props().as).toEqual('h1');
        });
    });
    describe('subtitle', () => {
        it('should render subtitle in SubHeader Title Element', () => {
            const wrapper = shallow(<Stepper {...props} subTitle="test subtitle"/>);
            expect(wrapper.find('HeaderSubTitle').children().text()).toEqual('test subtitle');
        });
        it('should render subtitle with h2 as a "as" prop', () => {
            const wrapper = shallow(<Stepper {...props} subTitle="test subtitle"/>);
            expect(wrapper.find('HeaderSubTitle').props().as).toEqual('h2');
        });
    });

    describe('step', () => {
        context('when step is more than 1', () => {
            it('should render the two button icon', () => {
                const wrapper = shallow(<Stepper {...props} step={2}/>);
                expect(wrapper.find(ButtonIcon).length).toEqual(2);
            });
        });
        context('when step is 1 or less', () => {
            it('should render the one button icon', () => {
                const wrapper = shallow(<Stepper {...props} step={1}/>);
                expect(wrapper.find(ButtonIcon).length).toEqual(1);
            });
        });
    });

    // describe('step and steps', () => {
    //     context('when step is less than steps', () => {
    //         it('should show the next button', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} steps={3}/>);
    //             expect(wrapper.find('ButtonStyle').children().at(0).text()).toEqual('Next');
    //         });
    //
    //         it('should show the next icon component inside the button', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} steps={3}/>);
    //             expect(wrapper.find('ButtonStyle').children().at(1).props().name).toEqual('chevron-right');
    //         });
    //     });
    //     context('when step is equal or greater than steps', () => {
    //         it('should show the Done button', () => {
    //             const wrapper = shallow(<Stepper {...props} step={4} steps={3}/>);
    //             expect(wrapper.find('ButtonStyle').children().text()).toEqual('Done');
    //         });
    //     });
    // });

    // describe('onStepChange', () => {
    //     // mock funtions
    //     describe('when clicking on next', () => {
    //         const clickFn = jest.fn();
    //         it('should call function once', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} steps={3} onStepChange={clickFn}/>);
    //             wrapper.find('ButtonStyle').simulate('click');
    //             expect(clickFn).toHaveBeenCalledTimes(1);
    //         });
    //         const event = {};
    //         it('should call the function and pass the argument next with event', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} steps={3} onStepChange={clickFn}/>);
    //             wrapper.find('ButtonStyle').simulate('click', event);
    //             expect(clickFn).toHaveBeenCalledWith('next', event);
    //         });
    //     });

    //     describe('when clicking on prev button', () => {
    //         const clickFn = jest.fn();
    //         it('should call function once', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} onStepChange={clickFn}/>);
    //             wrapper.find('ButtonIcon').at(1).simulate('click');
    //             expect(clickFn).toHaveBeenCalledTimes(1);
    //         });
    //         const event = {};
    //         it('should call the function and pass the argument back with event', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} onStepChange={clickFn}/>);
    //             wrapper.find('ButtonIcon').at(1).simulate('click', event);
    //             expect(clickFn).toHaveBeenCalledWith('back', event);
    //         });
    //     });
    //
    //     describe('when clicking on done button', () => {
    //         const clickFn = jest.fn();
    //         it('should call function once', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} steps={2} onStepChange={clickFn}/>);
    //             wrapper.find('ButtonStyle').simulate('click');
    //             expect(clickFn).toHaveBeenCalledTimes(1);
    //         });
    //         const event = {};
    //         it('should call the function and pass the argument next with event', () => {
    //             const wrapper = shallow(<Stepper {...props} step={2} steps={2} onStepChange={clickFn}/>);
    //             wrapper.find('ButtonStyle').simulate('click', event);
    //             expect(clickFn).toHaveBeenCalledWith('next', event);
    //         });
    //     });
    //
    // });

    describe('onClose', () => {
        const clickFn = jest.fn();
        it('should call function once   ', () => {
            const wrapper = shallow(<Stepper {...props} step={1} onClose={clickFn}/>);
            wrapper.find('ButtonIcon').simulate('click');
            expect(clickFn).toHaveBeenCalledTimes(1);
        });
    });

    it('show step status according to passed step and steps', () => {
        const wrapper = shallow(<Stepper {...props} step={2} steps={3}/>);
        expect(wrapper.find('span.step-status').text()).toEqual('STEP 2 / 3');
    });

    it('renders a Stpper snapshot', () => {
        const wrapper = shallowWithTheme(<Stepper {...props} />, theme);
        expect(wrapper).toMatchSnapshot();
    });
});
