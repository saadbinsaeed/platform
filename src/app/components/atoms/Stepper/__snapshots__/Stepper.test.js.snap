// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> renders a Stpper snapshot 1`] = `ShallowWrapper {}`;
